go-controller/pkg/util/subnet_annotations.go

```diff
@@ -37,82 +38,110 @@ const (
+		delete(subnetsMap, netName)
+	}
+
+	// if no host subnet left, just delete the host subnet annotation from node annotations.
+	if len(subnetsMap) == 0 {
+		delete(annotations, annotationName)
+		return nil
+	}
+
+	// Marshall all host subnets of all networks back to annotations.
+	subnetsStrMap := make(map[string][]string)
+	for n, subnets := range subnetsMap {
+		subnetsStr := make([]string, len(subnets))
+		for i, subnet := range subnets {
+			subnetsStr[i] = subnet.String()
 		}
-		bytes, err = json.Marshal(map[string][]string{
-			"default": defaultSubnetStrs,
-		})
+		subnetsStrMap[n] = subnetsStr
 	}
+	bytes, err = json.Marshal(subnetsStrMap)
 	if err != nil {
-		return nil, err
+		return err
 	}
-	return map[string]interface{}{
-		annotationName: string(bytes),
-	}, nil
+	annotations[annotationName] = string(bytes)
+	return nil
 }
 
 func setSubnetAnnotation(nodeAnnotator kube.Annotator, annotationName string, defaultSubnets []*net.IPNet) error {
-	annotation, err := createSubnetAnnotation(annotationName, defaultSubnets)
+	annotation := map[string]string{}
+	err := updateSubnetAnnotation(annotation, annotationName, types.DefaultNetworkName, defaultSubnets)
 	if err != nil {
 		return err
 	}
 	return nodeAnnotator.Set(annotationName, annotation[annotationName])
 }
 
-func parseSubnetAnnotation(node *kapi.Node, annotationName string) ([]*net.IPNet, error) {
-	annotation, ok := node.Annotations[annotationName]
+func parseSubnetAnnotation(nodeAnnotations map[string]string, annotationName string) (map[string][]*net.IPNet, error) {
+	if annotationName != ovnNodeSubnets {
+		return nil, fmt.Errorf("incorrect subnet annotation name")
+	}
+	annotation, ok := nodeAnnotations[annotationName]
 	if !ok {
-		return nil, newAnnotationNotSetError("node %q has no %q annotation", node.Name, annotationName)
+		return map[string][]*net.IPNet{}, nil
 	}
-
-	var subnets []string
+	subnetsStrMap := map[string][]string{}
 	subnetsDual := make(map[string][]string)
 	if err := json.Unmarshal([]byte(annotation), &subnetsDual); err == nil {
-		subnets, ok = subnetsDual["default"]
+		subnetsStrMap = subnetsDual
 	} else {
 		subnetsSingle := make(map[string]string)
 		if err := json.Unmarshal([]byte(annotation), &subnetsSingle); err != nil {
-			return nil, fmt.Errorf("could not parse %q annotation %q as either single-stack or dual-stack",
-				annotationName, annotation)
+			return nil, fmt.Errorf("could not parse %q annotation %q as either single-stack or dual-stack: %v",
+				annotationName, annotation, err)
+		}
+		for netName, v := range subnetsSingle {
+			subnetsStrMap[netName] = make([]string, 1)
+			subnetsStrMap[netName][0] = v
 		}
-		subnets = make([]string, 1)
-		subnets[0], ok = subnetsSingle["default"]
 	}
-	if !ok {
-		return nil, fmt.Errorf("%q annotation %q has no default network", annotationName, annotation)
+
+	if len(subnetsStrMap) == 0 {
+		return nil, fmt.Errorf("unexpected empty %s annotation", annotationName)
```

Comment:
in the case of that particular unit tests, where annoation value is "{\"default\":[]}", subnetStrMap would be map["default":[]], lenth is non-zero.

Reply:
Yes but why are we treating no annotation (line 98) differently than `"{}"` (here)? I guess it was like this before, so no strong opinion

Reply:
Checking if you have considered this.

Reply:
no annotation is considered error before, and I'd rather not change that behavior.

Reply:
You did change the behavior.

No annotation was considered an error before but not now:
```
	if !ok {
-		return nil, newAnnotationNotSetError("node %q has no %q annotation", node.Name, annotationName)
+		return map[string][]*net.IPNet{}, nil
	}
```

An empty annotation is considered an error, this situation was not evaluated before this PR, and I don't know why it should be treated any differently than the no annotation case.


An nonexistent (default) network in the annotation was considered an error before here, and still is now in `ParseNodeHostSubnetAnnotation` :
```
subnets, ok := subnetsMap[netName]
if !ok {
return nil, newAnnotationNotSetError("node %q has no %q annotation for network %s", node.Name, ovnNodeSubnets, netName)
}
```

Reply:
yes, it is still considered as an error, so the it returns failure in the export function. I didn't change that behavior. But I will fix both functions to avoid the confusion.